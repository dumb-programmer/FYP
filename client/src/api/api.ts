export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
