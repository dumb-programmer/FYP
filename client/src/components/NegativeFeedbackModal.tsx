import React from "react";
import { Action, Content, Header, Modal } from "./modal";
import NegativeFeedbackForm from "./NegativeFeedbackForm";

const formId = "negative-feedback-form";
export default function NegativeFeedbackModal({ dialogRef, onCancel }: { dialogRef: React.MutableRefObject<HTMLDialogElement | null>, onCancel: () => void }) {
    return <Modal dialogRef={dialogRef}>
        <Header>
            <h1 className="text-xl font-bold mb-8">Help Us Improve</h1>
        </Header>
        <Content>
            <NegativeFeedbackForm id={formId} />
        </Content>
        <Action>
            <button type="button" className="btn btn-ghost" onClick={onCancel}>Cancel</button>
            <button form={formId} type="submit" className="btn btn-primary">Submit</button>
        </Action>
    </Modal>;
}