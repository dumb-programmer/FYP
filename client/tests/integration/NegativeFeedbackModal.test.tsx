import React, { useRef } from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi } from "vitest";
import NegativeFeedbackModal from "../../src/components/NegativeFeedbackModal";
import userEvent from "@testing-library/user-event";
import { sendFeedback } from "../../src/api/api";

vi.mock("../../src/api/api", async () => ({ ...await vi.importActual("../../src/api/api"), sendFeedback: vi.fn() }))

const messageId = "abc";
const onSuccessMock = vi.fn();
const onCancelMock = vi.fn();

const WrapperComponent = () => {
    const ref = useRef<HTMLDialogElement | null>(null);

    return <NegativeFeedbackModal dialogRef={ref} messageId={messageId} onSuccess={onSuccessMock} onCancel={onCancelMock} />
}

const renderComponent = () => render(<WrapperComponent />);

describe("NegativeFeedbackModal", () => {
    it("renders correctly", () => {
        const { container } = renderComponent();

        expect(container).toMatchSnapshot();
    });

    it("modal is closed by default", () => {
        renderComponent();

        const dialog = screen.queryByRole("dialog");

        expect(dialog).not.toBeInTheDocument();
    });

    it("cancel button works", async () => {
        const { container } = renderComponent();

        const dialog = container.querySelector("dialog");

        dialog?.setAttribute("open", "");

        const cancelBtn = screen.getByRole("button", { name: /cancel/i });

        await userEvent.click(cancelBtn);

        expect(onCancelMock).toHaveBeenCalled();
    });

    it("form submission works", async () => {
        sendFeedback.mockReturnValue({ ok: true });

        const { container } = renderComponent();

        const dialog = container.querySelector("dialog");

        dialog?.setAttribute("open", "");

        await userEvent.type(screen.getByRole("textbox"), "hello");

        await userEvent.click(screen.getByRole("button", { name: /submit/i }));

        expect(sendFeedback).toHaveBeenCalled();
        expect(sendFeedback).toHaveBeenCalledWith({ category: "offensive/unsafe", comments: "hello", messageId, type: "negative" });
        expect(onSuccessMock).toHaveBeenCalled();
    });
});