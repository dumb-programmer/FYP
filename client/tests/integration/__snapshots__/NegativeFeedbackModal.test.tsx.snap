// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NegativeFeedbackModal > renders correctly 1`] = `
<div>
  <dialog
    class="modal"
    data-testid="modal"
    id=":r0:"
  >
    <div
      class="modal-box"
    >
      <h1
        class="text-xl font-bold mb-8"
      >
        Help Us Improve
      </h1>
      <form
        class="flex flex-col gap-6"
        id="negative-feedback-form"
      >
        <div
          class="form-control flex flex-wrap gap-4 flex-row text-sm"
        >
          <label
            class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all bg-slate-800 text-white hover:bg-slate-700"
          >
            <input
              class="radio radio-primary invisible absolute"
              name="category"
              type="radio"
              value="offensive/unsafe"
            />
            offensive/unsafe
          </label>
          <label
            class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all "
          >
            <input
              class="radio radio-primary invisible absolute"
              name="category"
              type="radio"
              value="not-factually-correct"
            />
            not factually correct
          </label>
          <label
            class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all "
          >
            <input
              class="radio radio-primary invisible absolute"
              name="category"
              type="radio"
              value="other"
            />
            other
          </label>
        </div>
        <textarea
          class="textarea input-bordered"
          name="comments"
          placeholder="(Optional) Something else you'd like to say"
        />
      </form>
      <div
        class="modal-action"
      >
        <button
          class="btn btn-ghost"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary"
          form="negative-feedback-form"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </dialog>
</div>
`;
